### PR TITLE
Refresh token support for OAuth/OIDC public clients

### PR DESCRIPTION
**Pull request by julienhuon**
Hello Misagh,

Third and last part of my work on the improvement of the OAuth2.0/OIDC support for public clients in CAS.

This PR adds support for refresh token for this kind of clients. Here the changelog:

* Add OAuth20RefreshToken Authenticator for OAuth/OIDC public clients 
only.

* Removed public clients support in the OAuth20ClientIDClientSecret 
Authenticator (worked with client_secret="" which is not specified in OAuth2.0/OIDC).

* Add tests for the new OAuth20RefreshToken Authenticator.

* Add test for the existing OAuth20ClientIdClientSecret Authenticator.

With this last PR public clients will be fully supported (PKCE, refresh token & revocation).

Please let me know if you need any further information.

Regards,
Julien

### CODE REVIEW

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java [hunk 1]
```diff
@@ -107,6 +108,8 @@ protected void validateCredentials(final UsernamePasswordCredentials credentials
      * 1. When the grant type is {@code password}, in which case the authentication will be performed by {@code OAuth20UsernamePasswordAuthenticator}
      * 2. When request contains OAuth {@code code} which was issued with a {@code code_challenge}, in which case the authentication will be
      * performed by {{@code OAuth20ProofKeyCodeExchangeAuthenticator}
+     * 3. When the grant type is {@refresh_token} and the request doesn't have any secret, in which case the authentication will be performed
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java:111]: Bad javadoc?

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java [hunk 2]
```diff
@@ -121,6 +124,14 @@ protected boolean canAuthenticate(final WebContext context) {
             return false;
         }
 
+        if (grantType.isPresent()
+            && OAuth20Utils.isGrantType(grantType.get(), OAuth20GrantTypes.REFRESH_TOKEN)
+            && context.getRequestParameter(OAuth20Constants.CLIENT_ID).isPresent()
+            && !context.getRequestParameter(OAuth20Constants.CLIENT_SECRET).isPresent()) {
+            LOGGER.debug("Skipping Client credential authentication to use refresh_token authentication");
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java:131]: Please correct the capitalization of the debug statement (i.e. Client), and possibly remove *refresh_token authentication*.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                                            final AuditableExecution registeredServiceAccessStrategyEnforcer,
+                                            final TicketRegistry ticketRegistry,
+                                            final CipherExecutor<Serializable, String> registeredServiceCipherExecutor,
+                                            final PrincipalResolver principalResolver) {
+        super(servicesManager, webApplicationServiceFactory, registeredServiceAccessStrategyEnforcer,
+            registeredServiceCipherExecutor, ticketRegistry, principalResolver);
+    }
+
+    /**
+     * RefreshTokenAuthenticator can only be used for a refresh token request of a "public" client.
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:44]: Bad javadoc. No such thing as `RefreshTokenAuthenticator`?

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+     * An OAuth "public" client is one that does not define a secret like a mobile application.
+     *
+     * @param context the context
+     * @return true if authenticator can validate credentials.
+     */
+    @Override
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:50]: Javadoc support for @Override methods is not strictly necessary. If you do need to explain this, you should do so at the class level.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    protected boolean canAuthenticate(final WebContext context) {
+        val grantType = context.getRequestParameter(OAuth20Constants.GRANT_TYPE);
+        val clientId = context.getRequestParameter(OAuth20Constants.CLIENT_ID);
+
+        if (clientId.isPresent() && grantType.isPresent()
+            && OAuth20Utils.isGrantType(grantType.get(), OAuth20GrantTypes.REFRESH_TOKEN)
+            && context.getRequestParameter(OAuth20Constants.REFRESH_TOKEN).isPresent()) {
+            val registeredService = OAuth20Utils.getRegisteredOAuthServiceByClientId(getServicesManager(), clientId.get());
+
+            if (registeredService != null && !OAuth20Utils.doesServiceNeedAuthentication(registeredService)) {
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:60]: Debug/Trace statements here would be very useful.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    protected void validateCredentials(final UsernamePasswordCredentials credentials,
+                                       final OAuthRegisteredService registeredService, final WebContext context) {
+        val token = credentials.getPassword();
+        LOGGER.trace("Received refresh token [{}] for authentication", token);
+
+        val refreshToken = getTicketRegistry().getTicket(token, OAuth20RefreshToken.class);
+        if (refreshToken == null || refreshToken.isExpired() || !StringUtils.equals(refreshToken.getClientId(), credentials.getUsername())) {
+            LOGGER.error("Provided refresh token [{}] is either not found in the ticket registry or has expired or not related with the client provided");
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:75]:
```suggestion
            LOGGER.error("Provided refresh token [{}] is either not found in the ticket registry, has expired or is not related to the client provided");
```

Comment (mmoayyed) [reply support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:75]: Could we also include the client-id in log statement?

Comment (julienhuon) [reply support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:75]: Sure, will add it.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java (hunk 2 above)
Comment (julienhuon) [reply support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java:131]: i chose to keep `refresh token authentication` part. It seems more logical with the previous statement isn't it ?

```
LOGGER.debug("Skipping Client credential authentication to use password authentication");
```

**Comment by julienhuon** [comment]
Just added a few tests to improve coverage...

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * OAuth20RefreshTokenAuthenticator can only be used for a refresh token request of a "public" client.
+     * An OAuth "public" client is one that does not define a secret like a mobile application.
+     *
+     * @param context the context
+     * @return true if authenticator can validate credentials.
+     */
+    protected boolean canAuthenticate(final WebContext context) {
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:50]: Missing `@Override`.

File: support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                                            final CipherExecutor<Serializable, String> registeredServiceCipherExecutor,
+                                            final PrincipalResolver principalResolver) {
+        super(servicesManager, webApplicationServiceFactory, registeredServiceAccessStrategyEnforcer,
+            registeredServiceCipherExecutor, ticketRegistry, principalResolver);
+    }
+
+    /**
+     * OAuth20RefreshTokenAuthenticator can only be used for a refresh token request of a "public" client.
+     * An OAuth "public" client is one that does not define a secret like a mobile application.
+     *
```
Comment (mmoayyed) [inline support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java:46]: Should be moved to the class.

File: support/cas-server-support-oauth/src/main/java/org/apereo/cas/config/CasOAuth20Configuration.java [hunk 9]
```diff
@@ -356,6 +364,18 @@ public OAuth20CallbackAuthorizeViewResolver callbackAuthorizeViewResolver() {
             defaultPrincipalResolver.getObject());
     }
 
+    @ConditionalOnMissingBean(name = "oAuthRefreshTokenAuthenticator")
+    @Bean
+    @RefreshScope
+    public Authenticator<UsernamePasswordCredentials> oAuthRefreshTokenAuthenticator() {
```
Comment (mmoayyed) [inline support/cas-server-support-oauth/src/main/java/org/apereo/cas/config/CasOAuth20Configuration.java:370]: `oauthRefreshTokenAuthenticator`, and same for the annotation.

Comment (julienhuon) [reply support/cas-server-support-oauth/src/main/java/org/apereo/cas/config/CasOAuth20Configuration.java:370]: Are you sure ? All the other authenticators have this syntax:
- oAuthProofKeyCodeExchangeAuthenticator
- oAuthClientAuthenticator
- oAuthAccessTokenAuthenticator
- etc...

**Comment by mmoayyed** [comment]
Excellent work. Thank you very much for the contribution. Keep 'em coming!

**Comment by julienhuon** [comment]
Thank you for the nice comment !

I still have a few fix for oauth/oidc in mind and maybe try this conformance test for oidc : https://openid.net/certification/testing/.

Could be nice to have CAS in the list of certified OIDC OP. :)

Regards,
Julien

**Comment by mmoayyed** [comment]
That work is in progress; we have a test instance here: https://op.certification.openid.net:61314/ running tests against a CAS instance that is deployed on Heroku, and about 75% of tests do pass. Unfortunately, due to lack of funding and commitment, the work has paused.

**Comment by julienhuon** [comment]
Will take a look at it ! Thanks.

**Pull request merged**